src/box/lua/config/utils/schema.lua

```diff
@@ -1445,7 +1445,79 @@ end
 
 -- {{{ <schema object>:merge()
 
+-- Are all keys in the given table strings?
+local function table_is_object(t)
+    assert(type(t) == 'table')
+
+    for k, _ in pairs(t) do
+        if type(k) ~= 'string' then
+            return false
+        end
+    end
+
+    return true
+end
+
+-- Perform a deep merge with some specific rules:
+--
+-- * Prefer box.NULL over nil.
+-- * Prefer X ~= nil over nil/box.NULL.
+-- * Prefer B if A or B is not a table.
+-- * Prefer B if A or B is a table with a non-string key.
+-- * Perform the deep merge otherwise.
+local function merge_schemaless(a, b)
+    -- Prefer box.NULL over nil.
+    if a == nil and b == nil then
+        if type(a) == 'nil' then
+            return b
+        else
+            return a
+        end
+    end
+
+    -- Prefer X ~= nil over nil/box.NULL.
+    if a == nil then
+        return b
+    elseif b == nil then
+        return a
+    end
+
+    assert(a ~= nil and b ~= nil)
+
+    -- Prefer B if A or B is not a table.
+    if type(a) ~= 'table' or type(b) ~= 'table' then
+        return b
+    end
+
+    -- Prefer B if A or B is a table with a non-string key.
+    if not table_is_object(a) or not table_is_object(b) then
+        return b
+    end
+
+    -- Perform the deep merge otherwise.
+    local res = {}
+
+    for k, v in pairs(a) do
+        res[k] = merge_schemaless(v, b[k])
+    end
+
+    for k, v in pairs(b) do
+        -- Process B's fields that are not present in A.
+        if type(a[k]) == 'nil' then
+            res[k] = merge_schemaless(a[k], v)
+        end
+    end
```

Comment:
Seems like replacing with this loop result with lack of `a` fields not present in `b` in `res` (I've tested it). Probably, this suggestion could be applied if `res` is initialized with `a` copy. Though the current implementation IMO is OK.

Reply:
OTOH, it actually has no sense to step into the recursion, when the left-hand side value is known as `nil`.

Reply:
I've adjusted it a bit.

```lua
diff --git a/src/box/lua/config/utils/schema.lua b/src/box/lua/config/utils/schema.lua
index dcdb0efa89..4ee4a7c625 100644
--- a/src/box/lua/config/utils/schema.lua
+++ b/src/box/lua/config/utils/schema.lua
@@ -1502,9 +1502,12 @@ local function merge_schemaless(a, b)
     end
 
     for k, v in pairs(b) do
-        -- Process B's fields that are not present in A.
+        -- Copy B's fields that are not present in A.
+        --
+        -- Don't do the recursive call here, because A's field is
+        -- known as nil.
         if type(a[k]) == 'nil' then
-            res[k] = merge_schemaless(a[k], v)
+            res[k] = v
         end
     end
 
```